### PR TITLE
Implement ECE for payment processing on block checkout page

### LINKED ISSUE #3403: [ECE] Implement ECE for payment processing on block checkout page
### Description
Process payment with the ECE button on the block checkout page.

### Acceptance Criteria
- ECE buttons are rendered on the block checkout page.
- You can successfully process payment with the ECE button from the block checkout page.
- Both success and failure cases are handled including necessary notices.

### PR DESCRIPTION
**Pull request by Mayisha**
Fixes #3403  #3408

## Changes proposed in this Pull Request:
- Created hooks, event-handlers and utils functions for ECE in blocks.
- The changes are made following the functions for [payment-request](https://github.com/woocommerce/woocommerce-gateway-stripe/tree/develop/client/blocks/payment-request) and WooPayments. 

## Testing instructions
- Enable the ECE feature flag (return true from `is_stripe_ece_enabled` function or in your database add an option `_wcstripe_feature_ece` and set yes as the value).
- As a shopper add a product to your cart and go to the Block checkout page.
- Confirm that the Google Pay and Apple Pay buttons are loaded.
- Confirm that you can successfully place an order with Google Pay and Apple Pay.
- As a shopper add a product to your cart and go to the Block cart page.
- Confirm that the Google Pay and Apple Pay buttons are loaded.
- Confirm that you can successfully place an order with Google Pay and Apple Pay.
- Disable the feature flag and confirm that PRBs are still working without any regression.

### CODE REVIEW

**Comment by Mayisha** [comment]
@wjrosa I have requested your review as you are also working on ECE issues. Feel free to unassign yourself if you are too occupied 🍫

File: client/blocks/express-checkout/express-checkout-component.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+		<ExpressCheckoutElement
+			options={ {
+				...buttonOptions,
+				...getPaymentMethodsOverride( expressPaymentMethod ),
+			} }
+			onClick={ onButtonClick }
+			onConfirm={ onConfirm }
+			onReady={ onElementsReady }
+			onCancel={ onCancel }
+			// onShippingAddressChange={ onShippingAddressChange }
```
Comment (wjrosa) [inline client/blocks/express-checkout/express-checkout-component.js]: Is this intended?

**Comment by annemirasol** [comment]
✨ Great work ✨  I can see Google Pay in my initial testing.

1. Is it expected that an unavailable component still takes up space? In my case, Apple Pay is not available (ie Chrome browser) but the container is still rendered.

_Empty Apple Pay container taking up space:_
Screenshot 2024-09-17 at 2 01 08 PM

2. When `is_stripe_ece_enabled` returns `false`, I get an empty Express Checkout box. I expected the old Google Pay PRB (or for the entire Express Checkout box to not show at all).

_Empty box when Stripe ECE is not enabled:_
Screenshot 2024-09-17 at 2 04 25 PM

_What I see on `develop`:_
Screenshot 2024-09-17 at 2 06 54 PM

File: client/blocks/express-checkout/express-checkout-component.js (hunk 1 above)
Comment (Mayisha) [reply client/blocks/express-checkout/express-checkout-component.js]: Yes for now. The shipping change is causing an issue which I need to fix.

**Comment by Mayisha** [comment]
Thanks, @annemirasol for checking this.

> Empty Apple Pay container taking up space

I will create a separate issue to handle this scenario. 👀 

> Empty box when Stripe ECE is not enabled:

After disabling the feature flag, try running `npm start` again and see if the PRB button loads. Probably the PRB script is not loaded and that's causing the issue. 🤔

**Comment by annemirasol** [comment]
> After disabling the feature flag, try running `npm start` again and see if the PRB button loads. Probably the PRB script is not loaded and that's causing the issue. 🤔

I'm still getting an empty checkbox after running `npm start` and doing a full reload. I see errors in the console:

Screenshot 2024-09-18 at 9 34 40 AM

(The hooks.js line with error)
Screenshot 2024-09-18 at 9 35 09 AM

(For comparison, this is my browser console on `develop`.)
Screenshot 2024-09-18 at 9 31 40 AM

File: includes/class-wc-stripe-blocks-support.php [hunk 2]
```diff
@@ -174,6 +183,7 @@ public function get_payment_method_data() {
 		return array_replace_recursive(
 			$this->get_gateway_javascript_params(),
 			$this->get_payment_request_javascript_params(),
+			$this->get_express_checkout_javascript_params(),
```
Comment (annemirasol) [inline includes/class-wc-stripe-blocks-support.php]: @Mayisha Had some time to closely look at the empty box/403 forbidden problem I reported earlier, and it looks like it's from this line ^ that's not checking whether ECE is enabled or not before replacing JS (nonce) params

Comment (Mayisha) [reply includes/class-wc-stripe-blocks-support.php]: Thanks @annemirasol for the findings 😃

**Comment by Mayisha** [comment]
@wjrosa @annemirasol this PR is ready for review again. The changes seems to work for block cart page as well. So please test this on both the block cart and block checkout pages.

There are 2 issues I am going to fix separately.
- Empty space for Apple Pay container that @annemirasol mentioned in [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3439#issuecomment-2356695099). (Issue #3470 )

Screenshot 2024-09-25 at 1 02 55 AM


- The payment method name is `Credit/Debit Card` instead of Google Pay or Apple Pay. This happens in the [shortcode PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3436) as well. So we will fix it later for both. (Issue #3469)

Screenshot 2024-09-25 at 12 57 24 AM

**Review by wjrosa — APPROVED** [review]
Great work, Mayisha! Code is good and it works as expected:

> Confirm that the Google Pay and Apple Pay buttons are loaded.

Screenshot 2024-09-25 at 11 56 13

> Confirm that you can successfully place an order with Google Pay and Apple Pay.

Screenshot 2024-09-25 at 11 59 42

> Confirm that the Google Pay and Apple Pay buttons are loaded.

Screenshot 2024-09-25 at 12 02 18

> Confirm that you can successfully place an order with Google Pay and Apple Pay.

Screenshot 2024-09-25 at 12 03 00

> Disable the feature flag and confirm that PRBs are still working without any regression.

Screenshot 2024-09-25 at 12 00 32

**Comment by annemirasol** [comment]
This could be a me thing, but I can't get Apple Pay to show up on this branch, when ECE is enabled.

How it looks on `develop`, or for this branch when `is_stripe_ece_enabled` == `false`:
Screenshot 2024-09-25 at 12 04 59 PM

How it looks for this branch, with `is_stripe_ece_enabled` == `true`:
Screenshot 2024-09-25 at 12 06 48 PM

I was able to test Google Pay, and these scenarios worked as expected ✨ 
- [x] Checkout: Google Pay displayed and working when `is_stripe_ece_enabled` is false
- [x] Cart: Google Pay displayed and working when `is_stripe_ece_enabled` is false
- [x] Checkout: Google Pay displayed and working when `is_stripe_ece_enabled` is true
- [x] Cart: Google Pay displayed and working when `is_stripe_ece_enabled` is true

Some other things I observed, which I think are out-of-scope for this PR, but noting here to document. LMK if you want me to file issues for these.
- When Apple Pay/Google Pay are not enabled as payment methods, and `is_stripe_ece_enabled` is`true`, I get empty boxes/sections.
- Noting that we may need to copy over the PRB fix for UAE addresses: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3432

Cart: Apple Pay/Google Pay are not enabled as payment methods, `is_stripe_ece_enabled` is `true`.
Screenshot 2024-09-25 at 11 25 41 AM

Checkout: Apple Pay/Google Pay are not enabled as payment methods, `is_stripe_ece_enabled` is `true`.
Screenshot 2024-09-25 at 11 25 52 AM

UAE addresses aren't being mapped correctly. This was recently fixed for PRBs:
Screenshot 2024-09-25 at 11 49 06 AM

File: client/blocks/express-checkout/index.js [hunk 3]
```diff
@@ -1,20 +1,44 @@
...
+	},
+} );
+
+const expressCheckoutElementsApplePay = ( api ) => ( {
+	name: PAYMENT_METHOD_EXPRESS_CHECKOUT_ELEMENT + '_applePay',
+	content: (
+		<ExpressCheckoutContainer
+			api={ api }
+			stripe={ stripePromise }
+			expressPaymentMethod="applelePay"
```
Comment (james-allan) [inline client/blocks/express-checkout/index.js]:
```suggestion
			expressPaymentMethod="applePay"
```

Typo. :)

**Comment by james-allan** [comment]
> This could be a me thing, but I can't get Apple Pay to show up on this branch, when ECE is enabled.

Not just a you thing I thing. I'm seeing the same thing. 
Screenshot 2024-09-26 at 1 25 31 pm

I noticed the following error in my console log: 

Screenshot 2024-09-26 at 1 24 43 pm

Fixing that (see [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3439/files#r1776304391)) will fix it for me. :)

Screenshot 2024-09-26 at 1 28 50 pm

File: client/blocks/express-checkout/index.js (hunk 3 above)
Comment (Mayisha) [reply client/blocks/express-checkout/index.js]: Thanks @james-allan 🤣 
Fixed in [7a21842](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3439/commits/7a218427e8fbec12cff6ee8f7732759cbd701913)

**Comment by Mayisha** [comment]
Thanks everyone for the review.

@annemirasol these are good findings.

> When Apple Pay/Google Pay are not enabled as payment methods, and is_stripe_ece_enabled istrue, I get empty boxes/sections.

I will fix it under https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3470

> Noting that we may need to copy over the PRB fix for UAE addresses: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3432

I have fixed this in [a1896eb](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3439/commits/a1896ebfd9b40944b2c87e1cdd07b86a9b7fab45) and noted it in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3412

**Pull request merged**